### PR TITLE
Replaced binary operators with intrinsics in AVX2 and SSSE3 codecs, proper check of CPU's AVX2 feature support

### PR DESCRIPTION
**Pull request by zenden2k**
I have created a new pull request to keep history clean.

### CODE REVIEW

**Comment by aklomp** [comment]
It's actually possible in GitHub to update a pull request by force-pushing over your branch. The pull request will be updated automatically.

File: lib/codec_choose.c [hunk 1]
```diff
@@ -90,18 +117,41 @@ codec_choose_arm (struct codec *codec)
...
 	if (max_level >= 7) {
 		__cpuid_count(7, 0, eax, ebx, ecx, edx);
-		if (ebx & bit_AVX2) {
-			codec->enc = base64_stream_encode_avx2;
-			codec->dec = base64_stream_decode_avx2;
+
+		if ((ebx & bit_AVX2) && (ecx & bit_XSAVE_XRSTORE)) {
+			uint64_t xcr_mask;
+			xcr_mask = _xgetbv_32(0);
+			if (xcrFeatureMask & 0x6) {
```
Comment (aklomp) [inline lib/codec_choose.c]: This causes a compilation error: `'xcrFeatureMask' undeclared`

Comment (zenden2k) [inline lib/codec_choose.c]: Fixed

**Comment by zenden2k** [comment]
Fixed and force-pushed

**Comment by zenden2k** [comment]
Travis CI can't check AVX2?

**Comment by aklomp** [comment]
No, the Travis virtual machines don't have AVX2 support. It might be possible to test AVX2 support by running the test suite under Intel SDE (an instruction set emulator).

Your changeset is starting to look good, and I will merge it when the last issues are worked out. (I might make some small changes to the codeflow later, but I will leave your contribution intact.) However, because it requires C99, I need to make some commits to this project to switch to C99 first. Working on it...

**Comment by aklomp** [comment]
OK, this library is now C99 :)  Now that we're C99, I changed the return type of the `codec_*()` functions to `bool`.

I also updated the Moby Dick tests to include the plain and encoded data as inline strings. So the funky compile-as-binary-object stuff is gone, and the test should work on Windows.

**Comment by zenden2k** [comment]
>No, the Travis virtual machines don't have AVX2 support.
Maybe it could at least check that compilation succeeds with AVX2 flag turned on.

**Comment by aklomp** [comment]
Merged, thanks!

**Pull request merged**